src/storage/nssdb/mod.rs

```diff
@@ -0,0 +1,1234 @@
+                        };
+                        let number = u32::from_be_bytes(bytes);
+                        let ulong = number as CK_ULONG;
+                        Attribute::from_attr_slice(
+                            cols[i],
+                            atype,
+                            &ulong.to_ne_bytes(),
+                        )
+                    }
+                    AttrType::BoolType
+                    | AttrType::StringType
+                    | AttrType::BytesType
+                    | AttrType::DateType => {
+                        if blob == &NSS_SPECIAL_NULL_VALUE {
+                            Attribute::from_attr_slice(cols[i], atype, &[])
+                        } else {
+                            Attribute::from_attr_slice(cols[i], atype, blob)
+                        }
+                    }
+                    AttrType::IgnoreType => {
+                        Attribute::from_attr_slice(cols[i], atype, &[])
+                    }
+                    AttrType::DenyType => {
+                        return Err(CKR_ATTRIBUTE_TYPE_INVALID)?
+                    }
+                };
+                obj.set_attr(attr)?;
+            }
+        }
+
+        /* ensure only one row was returned */
+        match rows.next() {
+            Ok(r) => match r {
+                Some(_) => Err(CKR_GENERAL_ERROR)?,
+                None => Ok(obj),
+            },
+            Err(_) => Err(CKR_GENERAL_ERROR)?,
+        }
+    }
+
+    fn prepare_fetch(
+        table: &str,
+        objid: u32,
+        attrs: &[CK_ATTRIBUTE],
+    ) -> Result<NSSSearchQuery> {
+        let mut columns: String;
+        if attrs.len() == 0 {
+            columns = "*".to_string();
+        } else {
+            columns = String::new();
+            for i in 0..attrs.len() {
+                if i == 0 {
+                    write!(&mut columns, "a{:x}", attrs[0].type_)?;
+                } else {
+                    write!(&mut columns, ", a{:x}", attrs[i].type_)?;
+                }
+            }
+        }
+        let mut query = NSSSearchQuery {
+            public: None,
+            private: None,
+            params: Vec::<Value>::with_capacity(1),
+        };
+        let sql = format!(
+            "SELECT DISTINCT {} FROM {} WHERE id = ? LIMIT 1",
+            columns, table
+        );
+        match table {
+            NSS_PUBLIC_TABLE => query.public = Some(sql),
+            NSS_PRIVATE_TABLE => query.private = Some(sql),
+            _ => return Err(CKR_GENERAL_ERROR)?,
+        }
+
+        query.params.push(Value::from(objid));
+
+        Ok(query)
+    }
+
+    fn fetch_by_nssid(
+        &self,
+        table: &str,
+        objid: u32,
+        attrs: &[CK_ATTRIBUTE],
+    ) -> Result<Object> {
+        let query = Self::prepare_fetch(table, objid, &attrs)?;
+        let sql = if let Some(ref s) = query.public {
+            s
+        } else if let Some(ref s) = query.private {
+            s
```

Comment:
unfortunately the table is not a variable that can be substituted in sql statement (at least it did not work when I tried to pass it as one of the params). Because for databases we may need to search two tables and therefore I have to create two queries I use th same construct also for single fetches, and then we need to find out wheher the public or the private table need to be searches bases on which sql statement has been set.